hyperapp.d.ts

```diff
@@ -51,58 +56,126 @@ export function h<Props>(
-}
+export type ActionResult<TState extends Primitive | NestedMap<Primitive>> =
+  | Partial<TState>
+  | Promise<any>
+  | null
+  | void
 
-/** The view function describes the application UI as a tree of VNodes.
+/** 
+ * The view function describes the application UI as a tree of VNodes.
  * @returns A VNode tree.
  * @memberOf [App]
  */
-export interface View<State, Actions> {
-  (state: State, actions: Actions): VNode<object>
+export interface View<
+  TState extends StateType,
+  TActions extends WiredActions<TState, TActions>
+> {
+  (state: TState, actions: TActions): VNode<any>
 }
 
-/** The app() call creates and renders a new application.
+/** 
+ * The app() call creates and renders a new application.
  *
  * @param state The state object.
  * @param actions The actions object implementation.
  * @param view The view function.
  * @param container The DOM element where the app will be rendered to.
  * @returns The actions wired to the application.
- * @memberOf [App]
  */
-export function app<State, Actions>(
-  state: State,
-  actions: ActionsType<State, Actions>,
-  view: View<State, Actions>,
-  container: Element | null
-): Actions
+export declare const app: App
 
-/** @namespace [JSX] */
+/**
+ * The interface for app(). Use this for implementing Higher Order App's
+ */
+export interface App {
+  <TState extends StateType, TActions extends UnwiredActions<TState, TActions>>(
+    state: TState,
+    actions: TActions,
+    view: View<TState, ActionProxy<TState, TActions>>,
+    container: HTMLElement | null
+  ): ActionProxy<TState, TActions>
+}
 
+/** 
+ * The type of action that's been returned from app()
+ * 
+ * This action is also passed as second parameter to the function that's 
+ * returned from unwired actions.
+ */
+type WiredAction<
+  TState extends Primitive | NestedMap<Primitive>,
+  TPayload extends Primitive | NestedMap<Primitive> = any
+> = (data?: TPayload) => TState
+
+/** 
+ * The action that's being passed as second param in an object to app(). 
+ * Use this when implementing your actions.
+ */
+export type UnwiredAction<
+  TState extends StateType,
+  TActions extends UnwiredActions<TState, TActions>,
+  TPayload extends Primitive | NestedMap<Primitive> = any
+> = (
+  data?: TPayload
+) =>
+  | ((
+      state: TState,
+      actions: ActionProxy<TState, TActions>
+    ) => ActionResult<TState>)
+  | ActionResult<TState>
+
+/** 
+ * Convenience type for defining initial State. 
+ * Extend your state object from this interface.
+ */
+export interface StateType extends NestedMap<Primitive> {}
+
+/** 
+ * Convenience type for defining your Actions.
+ * Extend your actions object from this interface.
+ */
+interface UnwiredActions<
+  TState extends StateType,
+  TActions extends UnwiredActions<TState, ActionProxy<TState, TActions>>
+> extends NestedMap<UnwiredAction<TState, ActionProxy<TState, TActions>>> {}
+
+/**
+ * Convenience type for NestedMap<WiredAction>
+ */
+interface WiredActions<
```

Comment:
Since you implement this. It would be nice to use `Actions` here. You can then use `ActionsTemplate` then for the `UnwiredActions`

Reply:
Another proposal: `Actions` and `AppActions`. Where `Actions` are the unwired ones and `AppActions` the wired ones

Reply:
It would be nice to use `Actions` for the most common one.

Reply:
What is 'most common' where you implement actions, or where you use them (e.g. in actions)?